pkg/lib/aws/servicequotas.go

```diff
@@ -103,3 +110,156 @@ func (c *Client) VerifyInstanceQuota(instanceType string, requiredOnDemandInstan
 
 	return nil
 }
+
+func (c *Client) VerifyNetworkQuotas(requiredInternetGateways int, requiredNATGatewaysPerAZ int, highlyAvailableNATGateway bool, requiredVPCs int, availabilityZones strset.Set) error {
+	ec2QuotaCodeToValueMap := map[string]int{
+		_elasticIPsQuotaCode: 0, // elastic IP quota code
+	}
+	vpcQuotaCodeToValueMap := map[string]int{
+		_internetGatewayQuotaCode: 0, // internet gw quota code
+		_natGatewayQuotaCode:      0, // nat gw quota code
+		_vpcQuotaCode:             0, // vpc quota code
+	}
+
+	err := c.ServiceQuotas().ListServiceQuotasPages(
+		&servicequotas.ListServiceQuotasInput{
+			ServiceCode: aws.String("ec2"),
+		},
+		func(page *servicequotas.ListServiceQuotasOutput, lastPage bool) bool {
+			if page == nil {
+				return false
+			}
+			for _, quota := range page.Quotas {
+				if quota == nil || quota.QuotaCode == nil || quota.Value == nil {
+					continue
+				}
+				if _, ok := ec2QuotaCodeToValueMap[*quota.QuotaCode]; ok {
+					ec2QuotaCodeToValueMap[*quota.QuotaCode] = int(*quota.Value)
+					return false
+				}
+			}
+			return true
+		},
+	)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	err = c.ServiceQuotas().ListServiceQuotasPages(
+		&servicequotas.ListServiceQuotasInput{
+			ServiceCode: aws.String("vpc"),
+		},
+		func(page *servicequotas.ListServiceQuotasOutput, lastPage bool) bool {
+			if page == nil {
+				return false
+			}
+			for _, quota := range page.Quotas {
+				if quota == nil || quota.QuotaCode == nil || quota.Value == nil {
+					continue
+				}
+				if _, ok := vpcQuotaCodeToValueMap[*quota.QuotaCode]; ok {
+					vpcQuotaCodeToValueMap[*quota.QuotaCode] = int(*quota.Value)
+				}
+			}
+			return true
+		},
+	)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// get EIP in use
+	elasticIPsInUse, err := c.ListElasticIPs()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// get IGW in use
+	internetGatewaysInUse, err := c.ListInternetGateways()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// get NAT GW in use per selected AZ
+	gateways, err := c.DescribeNATGateways()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	subnets, err := c.DescribeSubnets()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	gatewaysSubnetIDs := SubnetIDsFromNATGateways(gateways)
```

Comment:
nit: I think this code might be easier to follow if the `SubnetIDsFromNATGateways()` helper is removed, and the for loop below loops over `natGateways` instead of `gatewaysSubnetIDs` (and the subnet ID is extracted inside of the for loop)

Reply:
I agree!